human readable errors for insomnia config validation

dimitropoulos opened the pull request:
closes INS-1032

This PR introduces the following quality-of-life improvement to insomnia config errors:

| BEFORE | AFTER |
| - | - |
| ![Screenshot_20211019_171720](https://user-images.githubusercontent.com/15232461/138016713-ddfed116-0ce8-4054-b2a4-354e2f4dc16d.png) | ![Screenshot_20211019_212250](https://user-images.githubusercontent.com/15232461/138016777-dbf8ee75-c3b4-4e59-9294-78e83dc22763.png) |

dimitropoulos committed:
better errors

dimitropoulos committed:
rework

develohpanda commented on packages/insomnia-app/app/models/helpers/settings.ts:
```diff
@@ -71,7 +71,8 @@ interface ConfigError {
   error: {
     configPath?: string;
     insomniaConfig: unknown;
```
I've noticed this a couple of times and am not asking for any changes here but just want to understand, why is it that this needs to be typed as `unknown`?

dimitropoulos replied:
Think about it this way: if there's an error, then by definition the thing provided is _not_ an `InsomniaConfig`.  But what is it?  Well, we don't know.  Hence, `unknown`.  We can't make any assumptions about the shape of this object, so we just have to take it piece by piece (if we need pieces of the data), which is what `unknown` requires you to do.

develohpanda commented on packages/insomnia-config/src/validate.ts:
```diff
@@ -22,12 +23,13 @@ export const ingest = (input: string | InsomniaConfig | unknown) => {
   valid: true;
   errors: null;
-  humanError: null;
+  humanErrors: null;
```
Why `null` instead of omitting the property (both `errors` and `humanErrors`) altogether?

dimitropoulos replied:
the reason for this approach is to avoid having to make typeguards, so that (ideally) typechecking will automatically flow when you do a membership check (i.e. with the `in` operator or something).  now that there's a typeguard in your recent commit, then, yes, it's not needed anymore.  removed in f80d69150108b8e6d9f6ed873f9b0d79084bc177

dimitropoulos replied:
![Screenshot_20211020_113250](https://user-images.githubusercontent.com/15232461/138124358-ad65ab58-4195-40a5-914a-83f00931c2e8.png)

Ok, actually, I think I take it back, and for the same reason I mentioned above.  I thought that it wouldn't be that hard in the code, and, indeed, it's not _that_ hard, but it makes testing much much more of a burden.

dimitropoulos replied:
added back in 9dfd8354573f1741006805f6bbebe5b883ac7789

develohpanda committed:
introduce type predicates to avoid as casting

develohpanda reviewed (dismissed):
Error on macOS, LGTM

![image](https://user-images.githubusercontent.com/4312346/138031535-651e5df8-a092-4be8-b06f-06b1234ead3a.png)

In the `[Error 1]` section, I see `.settings/insomnia.config.json`. I don't understand what that syntax means 🤔

develohpanda committed:
Update settings.ts

dimitropoulos committed:
removes null values now that guards exist

dimitropoulos committed:
add filename in one place, only

before this, the config file was being added in multiple places, and that created a scenario where (on line 33 of settings.ts, for example) one was missed.  To fix this, we can just add the file path at the beginning, and then never need to fuss with it again in subsequent usages and logging and error reporting.

dimitropoulos committed:
updates tests for humanErrors

gatzjames commented on packages/insomnia-config/src/validate.ts:
```diff
@@ -21,17 +22,19 @@ export const ingest = (input: string | InsomniaConfig | unknown) => {
 export interface ErrorResult {
   valid: false;
   errors: ErrorObject[];
+  humanErrors: ValidationError[];
```
This name leads me to think these are errors made by humans/users.
humanReadableErrors or validationErrors would be more precise imo

dimitropoulos replied:
ok, I see where you're coming from, I'll update now

dimitropoulos replied:
updated in ad398ee033f12c5019391f942108751371933774

dimitropoulos committed:
adds back nulls due to property membership checks

dimitropoulos committed:
fixes return of mocks for file path inclusion

gatzjames reviewed (commented):
~~Having an invalid json in the `insomniaDataDirectory` path still opens the app for me:~~ Addressed in another PR

If I provide an incorrect value I see an error as expected:
<img width="288" alt="insomnia-config-error" src="https://user-images.githubusercontent.com/12115431/138129593-e3383c81-c6d9-463c-a0d4-ce95f28ab069.png">

dimitropoulos committed:
updates test to be windows path resistant

gatzjames commented on packages/insomnia-app/app/models/helpers/settings.ts:
```diff
@@ -39,6 +38,10 @@ export const getLocalDevConfigFilePath = () => (
   isDevelopment() ? '../../packages/insomnia-app/app' as string : undefined
```
Not related to this PR but should we use the path util instead:
```suggestion
  isDevelopment() ? path.join(__dirname, '..', '..', 'packages', 'insomnia-app', 'app') as string : undefined
```

dimitropoulos replied:
I think it's related _enough_ since this literal path appears on the screen to users as part of the change to the modal made in this PR.  it's not going to ever fail since it goes through `resolve`, but yes, you're right, this should be fixed.

dimitropoulos replied:
update, heh.  it didn't work.. not sure why.  but yes, I agree, we can handle later

dimitropoulos committed:
`humanErrors` -> `humanReadableErrors`

gatzjames reviewed (approved):
🚀 LGTM
